### PR TITLE
NFC: GDP documentation update

### PR DESCRIPTION
**Pull request by qtothec**
## Summary/Motivation:
Documentation update for Pyomo.GDP. Includes formatting changes and describes TransformationFactory and GDPopt.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: doc/OnlineDocs/modeling_extensions/gdp.rst [hunk 1]
```diff
@@ -61,32 +60,36 @@ the example returns.
 Transformation
 --------------
 
-In order to use the solvers currently available, one must convert the
-disjunctive model to a standard MIP/MINLP model.  The easiest way to do
-that is using the (included) BigM or Convex Hull transformations.  From
-the Pyomo command line, include the option ``--transform
-pyomo.gdp.bigm`` or ``--transform pyomo.gdp.chull``
+To use standard commercial solvers, you must convert the disjunctive model to a standard MIP/MINLP model.
+The two classical strategies for doing so are the (included) Big-M and Hull reformulations.
```
Comment (blnicho) [inline doc/OnlineDocs/modeling_extensions/gdp.rst]: Is there a reason you're calling it the Hull reformulation instead of Convex Hull? The transformation is invoked using `chull` which might be a little confusing if you just refer to it as the Hull reformulation.

Comment (qtothec) [reply doc/OnlineDocs/modeling_extensions/gdp.rst]: Yes, because in literature it appears as the Hull Reformulation. Practically speaking, it is possible to apply it to nonconvex functions, although the theory around this is a bit sparse.

Comment (bernalde) [reply doc/OnlineDocs/modeling_extensions/gdp.rst]: This addresses an issue that might be a little imprecision from the notation in Pyomo. The reformulation is named the hull reformulation because it only takes the convex hull from the elements in the disjunction (if well written) but does not yield the actual object popularly known as convex hull from the object (i.e. if we forget about the integrality condition we wouldn't necessarily get an integer solution).

Comment (emma58) [reply doc/OnlineDocs/modeling_extensions/gdp.rst]: I agree that the name may be misleading, but just to note: It is the true convex hull, but represented in a higher-dimensional space.

File: doc/OnlineDocs/modeling_extensions/gdp.rst [hunk 2]
```diff
@@ -61,32 +60,36 @@ the example returns.
...
 
-Notes
------
+- ``TransformationFactory('gdp.bigm').apply_to(model)``
+- ``TransformationFactory('gdp.chull').apply_to(model)``
+
+.. note::
 
-Some notes:
+    - all variables that appear in disjuncts need upper and lower bounds
```
Comment (jsiirola) [inline doc/OnlineDocs/modeling_extensions/gdp.rst]: This is only true for thc HUll reformulation.  BigM only needs bounds if you are going to let us automatically determine M

Comment (bernalde) [reply doc/OnlineDocs/modeling_extensions/gdp.rst]: Even for nonlinear constraints?

File: doc/OnlineDocs/modeling_extensions/gdp.rst [hunk 3]
```diff
@@ -61,32 +60,36 @@ the example returns.
...
 
-- for all other models, you will need to provide the M values through a
-  “BigM” Suffix.
+    - When you declare a Disjunct, it (at declaration time) will automatically
+      have a variable “indicator_var” defined and attached to it.
+      After that, it is just a Var like any other Var.
 
-- the convex hull reformulation is only valid for linear and convex
-  nonlinear problems.  Nonconvex problems are not supported (and are not
-  checked for).
```
Comment (jsiirola) [inline doc/OnlineDocs/modeling_extensions/gdp.rst]: Why are you removing this warning?

File: doc/OnlineDocs/modeling_extensions/gdp.rst (hunk 2 above)
Comment (jsiirola) [reply doc/OnlineDocs/modeling_extensions/gdp.rst]: Correct.  BigM will happily relax the constraints without regard to the validity of the relaxation (if you provide an M value, we trust it is correct).

File: doc/OnlineDocs/modeling_extensions/gdp.rst (hunk 3 above)
Comment (bernalde) [reply doc/OnlineDocs/modeling_extensions/gdp.rst]: I might have some results here that I would be happy to talk to you about. Even in that case, I would just leave the warning (this might go back to the whole convex hull vs. hull reformulation debate)

File: doc/OnlineDocs/modeling_extensions/gdp.rst (hunk 1 above)
Comment (qtothec) [reply doc/OnlineDocs/modeling_extensions/gdp.rst]: Ok folks, I just checked with Ignacio yesterday. The Hull Reformulation is an exact reformulation from GDP into MINLP at the optimal solution, irregardless of convexity for the original GDP model constraints. The main complicating element of HR is the perspective function, but the function value is exact at indicator variable values of `y=1` and `y=0`. Therefore, its function value is exact for any feasible solutions of the GDP.

tldr: the perspective function of a nonconvex function will still be nonconvex, so the hull reformulation is still valid, but is not convex in the case of a GDP with nonconvex constraints.

Comment (blnicho) [reply doc/OnlineDocs/modeling_extensions/gdp.rst]: @qtothec so is this ready to merge?

Comment (qtothec) [reply doc/OnlineDocs/modeling_extensions/gdp.rst]: @blnicho Let me make a couple extra edits. I'll rerequest your review when I'm done.
